cluster_link: admin api definition

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

Introduces the protobuf definition for cluster linking API

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
bzl: Add googleapis for protobufs

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
proto/admin: Add shadow_link RPC service

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,438 @@
+    ClusterLink cluster_link = 1;
+
+    // The list of fields to update
+    // google.protobuf.FieldMask update_mask = 2;
```
I think we would need to implement this in our code gen

rockwotj commented on proto/redpanda/core/admin/BUILD:
```diff
@@ -11,8 +11,24 @@ proto_library(
-    protos = [":admin_proto"],
+    protos = [
+        ":admin_proto",
+        ":cluster_link_proto",
```
Let's split into another `redpanda_proto_library`. `redpanda_proto_library` should be 1:1 with `proto_library`

michael-redpanda replied:
ah ok!

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,438 @@
+        (google.api.field_behavior) = OUTPUT_ONLY,
+        (google.api.field_info).format = UUID4
+    ];
+    // string uid = 2;
```
done

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,438 @@
+/// Options for syncing security settings
+message SecuritySettingsSyncOptions {
+    /// Sync interval
+    int32 interval_ms = 1;
```
bad numbering

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,437 @@
+    /// The cluster link to update
+    ClusterLink cluster_link = 1;
+
+    // The list of fields to update
```
Leaving this in here for now

michael-redpanda committed:
proto/admin: Added common proto library

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,396 @@
+    /// The cluster link to update
+    ClusterLink cluster_link = 1;
+
+    // The list of fields to update
+    // google.protobuf.FieldMask update_mask = 2;
```
No

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,396 @@
+    /// The cluster link to update
+    ClusterLink cluster_link = 1;
+
+    // The list of fields to update
+    // google.protobuf.FieldMask update_mask = 2;
```
No

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,396 @@
+    ClusterLink cluster_link = 1;
+
+    // The list of fields to update
+    // google.protobuf.FieldMask update_mask = 2;
```
No

rockwotj replied:
You can add field mask with `@protobuf//google:field_mask_proto`

rockwotj replied:
We do need to specialize the field mask proto codegen tho

michael-redpanda committed:
admin: Wire in shadow_link_service into admin API

Signed-off-by: Michael Boquard <michael@redpanda.com>

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+            authz: SUPERUSER
+        };
+    }
+    rpc FailOverClusterLink(FailOverClusterLinkRequest) returns (ClusterLink) {
```
non CRUD needs a comment

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    ClusterLink cluster_link = 1;
+
+    /// The list of fields to update
+    optional google.protobuf.FieldMask update_mask = 2;
```
generally we don't make this optional in the field presence case, but ignore it if the `paths` vector is empty. I'm fine to keep this, but we want to standardize going forward. Is there a difference between an empty vector vs not set?

michael-redpanda replied:
[AIP-134](https://google.aip.dev/134) states:

> The field must be optional, and the service must treat an omitted field mask as an implied field mask equivalent to all fields that are populated (have a non-empty value).

Was trying to comply with this but maybe I misread it.

rockwotj replied:
It's optional in the field_behavior sense. This was written before `optional` was added back into `proto3`

michael-redpanda replied:
Ahhhhhhh see that bit of context was missing :)

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// Configuration for the internal kafka client
+    ClusterLinkClientOptions client_options = 1;
+    /// Topic metadata sync options
+    /// If not provided, defaults to enabled with no filters
```
`defaults to enabled with no filters` reads a bit funny to me. Maybe: `by default all data is fetched from the source cluster`.

michael-redpanda replied:
i can reword this, sure

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// Fetch request timeout
+    /// Valid is [1, ...]
+    /// If 0 is provided, defaults to 100ms
+    int32 fetch_wait_max_ms = 8;
```
should this be a Duration type?

michael-redpanda replied:
Yeah I went back and forth in my head.  I was trying to expose kafka options in a "kafka like" sort of way, but that's not a good reason.  I can switch these over to being duraiton.

rockwotj replied:
I think because it's not more fine grained than millis, it's OK if you want to keep it like this (ie you can't specify 100 micro max wait).

michael-redpanda replied:
leaving as is then

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    google.protobuf.Duration interval = 1;
+    /// The topic filters to use
+    repeated NameFilter topic_filters = 2;
+    /// The mirror properties to use.  If not provided will automatically
+    /// mirror:
```
how do you mirror no properties?

michael-redpanda replied:
Set `mirror_topic_properties_provideed` to `true` and leave `mirrored_topic_properties` empty.

michael-redpanda replied:
removing this per comment below

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// Indicates if the mirrored_topic_properties above was provided
+    /// If false, defaults to the above list
+    bool mirrored_topic_properties_provided = 4;
+    bool auto_topic_create_enabled = 5;
```
needs docs? I honestly still don't understand this property and don't know why someone would set it false.

michael-redpanda replied:
Yeah for this release we can remove it.  In future use cases when users can manually create mirror topics, they may not want this task running.

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    repeated string mirrored_topic_properties = 3;
+    /// Indicates if the mirrored_topic_properties above was provided
+    /// If false, defaults to the above list
+    bool mirrored_topic_properties_provided = 4;
```
Is this from aip.dev?

michael-redpanda replied:
Yeah i'm happy to pull this and the above thing out.  Originally written this way before I had a better understanding of how this API style was supposed to be.

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// If false, defaults to the above list
+    bool mirrored_topic_properties_provided = 4;
+    bool auto_topic_create_enabled = 5;
+    bool topic_properties_sync_enabled = 6;
```
What does this mean?

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+}
+
+/* Resources */
+/// A ClusterLink resource
```
protobuf comments should only be `//`

michael-redpanda replied:
booooo

rockwotj replied:
https://giphy.com/gifs/filmeditor-the-princess-bride-l2YWqU7ev0l5nfYTC

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+/// TLS file settings
+message TLSFileSettings {
+    /// Path to the CA
+    string ca = 1;
```
should these be suffixed with `_path`?

michael-redpanda replied:
sure

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// Indicates that the key has been set
+    bool key_set = 3 [(google.api.field_behavior) = OUTPUT_ONLY];
+    /// The cert
+    optional string cert = 4;
```
should this be of type `bytes`? Same with key.

michael-redpanda replied:
It's coming in PEM encoded so they would be strings

rockwotj replied:
Cool, we do strictly enforce strings are less than 128KiB

michael-redpanda replied:
I think the iceberg config uses PEM encoded strings, right?  Any issues with length there?

rockwotj replied:
Probably not

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,416 @@
+    /// The broker the task is running on
+    int32 broker_id = 4;
+    /// The shard the broker is running on
+    int32 shard = 5;
```
Do we need to expose the shard or should this be saved for a debug endpoint?

michael-redpanda replied:
Yeah maybe it's not necessary... depends on the design of CO offset migration.

rockwotj commented on proto/redpanda/core/admin/common/acl.proto:
```diff
@@ -0,0 +1,63 @@
+
+syntax = "proto3";
+
+package redpanda.core.admin.common;
```
Are this base enough to `redpanda` that we should put this in `redpanda.core` or `redpanda.core.common`?

michael-redpanda replied:
oh good point, was thinking `redpanda.core.admin` was the base package

rockwotj replied:
ping on this one

michael-redpanda replied:
yeah I can move it

rockwotj reviewed (commented):
Looks good, some nits and small questions

rockwotj commented on src/v/redpanda/admin/services/cluster_link.cc:
```diff
@@ -0,0 +1,50 @@
+ss::future<proto::admin::cluster_link>
+cluster_link_service_impl::create_cluster_link(
+  proto::admin::create_cluster_link_request) {
+    return ss::make_ready_future<proto::admin::cluster_link>();
```
could also stub with an unimplemented exception

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,399 @@
+    // Path to the key
+    optional string key_path = 2;
+    // Path to the cert
+    optional string cert_path = 3;
```
I don't think you need `optional` but can just treat empty string the same as not set

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,399 @@
+    string ca = 1;
+    // Key and Cert are optional but if one is provided, then both must be
+    // The key
+    optional string key = 2 [(google.api.field_behavior) = INPUT_ONLY];
```
I don't think you need optional for this or the cert, but just treat empty string as unset

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,399 @@
+    // The host to match.  If not set, will default to match all hosts
+    // with the specified `operation` and `permission_type`. Note that
+    // the asterisk `*` is literal and matches hosts that are set to `*`
+    optional string host = 4;
```
ditto note on empty string

rockwotj reviewed (commented):
Generally looks good to me, this is my last round of nits. Generally if possible it's prefered to not add `optional` unless you really need it. Do you want to move this out of draft and merge it?

michael-redpanda commented:
> Generally looks good to me, this is my last round of nits. Generally if possible it's prefered to not add optional unless you really need it. Do you want to move this out of draft and merge it?

Thanks... I was going to leave it as is for the time being until I got some feedback on the RFC

michael-redpanda commented:
@rockwotj I still think I want to keep the `optional` for `tls_settings` and `authentication_config` within `ClusterLinkClientOptions` so it's explicit whether or not there's a desire to set either of those options.

rockwotj commented:
> @rockwotj I still think I want to keep the `optional` for `tls_settings` and `authentication_config` within `ClusterLinkClientOptions` so it's explicit whether or not there's a desire to set either of those options.

That's fine, I was thinking more around the primitive values.

michael-redpanda commented:
Force push:

* Rebase off dev (sorry should have been a separate force push)
* Updated `FalOverClusterLink` to `FailOver` to permit failing over a specific topic or the entire link

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,404 @@
+            type: "redpanda.core.admin.ClusterLinkService/ClusterLink"
+        }
+    ];
+    // The name of the mirror topic to fail over, if not set will fail over
+    // the entire cluster link
+    string mirror_topic_name = 2 [(google.api.field_behavior) = OPTIONAL];
```
Should this be repeated?

michael-redpanda replied:
Was thinking that for a specific topic failover it would happen one at a time.

rockwotj replied:
I guess that makes it explicit it is not atomic... Yeah I'm good with this

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,404 @@
+        [(google.api.field_behavior) = REQUIRED];
+    // The Client ID
+    string client_id = 2 [(google.api.field_behavior) = OUTPUT_ONLY];
+    // If provided, this is the expected ID of the source cluster.  If it does
+    // not match then the connection will be rejected
+    string source_cluster_id = 3;
```
What is this?

rockwotj replied:
Where do I get an ID from a source cluster?

michael-redpanda replied:
Ah yeah I can expand that comment - it's effectively the `cluster.id` that's returned in the metadata response.  It's optional but provides a way of ensuring you're connecting to the cluster you want to.

michael-redpanda replied:
updated comment

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,404 @@
+    // SCRAM username
+    string username = 1;
+    // Password
+    string password = 2 [(google.api.field_behavior) = INPUT_ONLY];
```
Does this need a `password_set` like above?

michael-redpanda replied:
yeah probably not a bad idea

michael-redpanda replied:
Added

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,407 @@
+    string client_id = 2 [(google.api.field_behavior) = OUTPUT_ONLY];
+    // If provided, this is the expected ID of the source cluster.  If it does
+    // not match then the connection will be rejected.  If provided, this value
+    // must match the `ClusterId` field returned in the metadata response
```
nit: which metadata response? :smile:

michael-redpanda replied:
all the metadata responses 😛 

i'll update it 😉

rockwotj commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,407 @@
+    // The bootstrap servers to use
+    repeated string bootstrap_servers = 1
+        [(google.api.field_behavior) = REQUIRED];
+    // The Client ID
```
```suggestion
    // The Client ID for the Kafka RPC requests sent by this cluster link to the source cluster. 
```

rockwotj replied:
While obvious to us, mere mortals who configure this stuff may not be in the same headspace :)

rockwotj reviewed (dismissed):
two nitters for you

michael-redpanda committed:
pbgen: Add missing return for variant visit

Also removed r-value reference return

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Moved cluster_link to own directory in `admin/services/cluster_link`
* Added fix for pbgen and variants

michael-redpanda commented:
Force push:

* Moved `cluster_link` to own directory in `admin/services/cluster_link`
* https://github.com/redpanda-data/redpanda/pull/27039/commits/9111fe78f2bae025bd50dde22e0d9ac6a03d1bce <-- fixed issue with variant visit

rockwotj commented on src/v/redpanda/application.cc:
```diff
@@ -134,6 +134,7 @@
 #include "raft/group_manager.h"
 #include "raft/service.h"
 #include "redpanda/admin/server.h"
+#include "redpanda/admin/services/cluster_link.h"
```
needs updating

michael-redpanda replied:
ah boo

michael-redpanda commented:
Force push:

* Fixed wrong includes

michael-redpanda commented:
Force push:

Fixed failing test

oleiman reviewed (dismissed):
cool

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--93ad5edd-0265-44f9-a2ca-7ff406deb8f6-->
<!--TEST RESULTS-->
<!--6cede1bc-8e08-4f61-a5be-e5caa7115e34-->
<!--TEST RESULTS-->
<!--f07a384f-33e7-4d2f-91b5-4289eb49b9f6-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70338">build#70338</a></summary>


| test_class                 | test_method                    | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:---------------------------|:-------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ConsumerGroupBalancingTest | test_coordinator_nodes_balance | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70338#01988162-8e5d-4bd0-b578-fafa479fe31f | FLAKY         | 20/21    | upstream reliability is '98.7864077669903'. current run reliability is '95.23809523809523'. drift is 3.54831 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70400">build#70400</a></summary>


| test_class                | test_method   | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                |
|:--------------------------|:--------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write    | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70400#01988588-704a-40b0-b51c-623c9e4b3f94 | FLAKY         | 18/21    | upstream reliability is '94.0934065934066'. current run reliability is '85.71428571428571'. drift is 8.37912 and the allowed drift is set to 50. The test should PASS |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/70448">build#70448</a></summary>


| test_class                | test_method                          | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:-------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTxTest | test_write                           | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70448#0198898a-d144-4320-9d3c-acc56c2fa517 | FLAKY         | 20/21    | upstream reliability is '93.96551724137932'. current run reliability is '95.23809523809523'. drift is -1.27258 and the allowed drift is set to 50. The test should PASS |
| DataMigrationsApiTest     | test_creating_and_listing_migrations | null             | integration | https://buildkite.com/redpanda/redpanda/builds/70448#01988987-92ca-4a7f-99b7-a93a2faf3130 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             |
</details>
<!--/body-->

michael-redpanda commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+    // The topic filters to use
+    repeated NameFilter topic_filters = 2;
+    // Additional topic properties to mirror
+    // Replication factor, `max.message.bytes`, `cleanup.policy` and
```
s/REplication factor/Partition count

michael-redpanda commented on proto/redpanda/core/common/acl.proto:
```diff
@@ -0,0 +1,63 @@
+
+option (redpanda.pbgen.cpp_namespace) = "proto::admin";
+
+/// The ACL resource types
+enum ACLResource {
+    ACL_RESOURCE_UNSPECIFIED = 0;
+    ACL_RESOURCE_ANY = 1;
+    ACL_RESOURCE_CLUSTER = 2;
+    ACL_RESOURCE_GROUP = 3;
+    ACL_RESOURCE_TOPIC = 4;
+    ACL_RESOURCE_TXN_ID = 5;
+}
```
Missing SR ACLs

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+// Copyright 2025 Redpanda Data, Inc.
```
Have you considered versioning the protos?
In cloud were' following https://google.aip.dev/185, which suggests to have an explicit major version in the protos path.
i.e. `proto/redpanda/core/admin/v1/cluster_link.proto`, or `proto/redpanda/core/admin/v1alpha1/cluster_link.proto`.

This helps with deprecations, and experimentation with unstable APIs.

michael-redpanda replied:
@rockwotj any thoughts on this?

rockwotj replied:
I thought about it, I'm fine to put stuff in a `v2` namespace. We're not going to rev the protocol ever really however because of our compatibility story is not quite the same as "normal" APIs.

rockwotj replied:
I'm fine to do this, but let's do it in a followup PR. Not this one

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+option (redpanda.pbgen.cpp_namespace) = "proto::admin";
+
+service ClusterLinkService {
+    rpc CreateClusterLink(CreateClusterLinkRequest) returns (ClusterLink) {
```
The convention for responses is to always the `Response` suffix to the method name. i.e. `CreateClusterLinkResponse`

simon0191 replied:
The buf linter and formatter is quite helpful with stuff like this

https://buf.build/docs/lint/

rockwotj replied:
That is against https://google.aip.dev/121#methods

michael-redpanda replied:
We're following the style of [AIP-133](https://google.aip.dev/133) here where things like `Create` or `Get` just return the resource

birdayz replied:
cloud chose to side with buf, not with AIP on this one:
- Consistency: there's a wrapper CreateX for requests, but not for responses
- You're locked out of having extra fields outside the resource
- This looks especially ugly if you look at JSON style, especially over REST (we use gRPC gateway atm on top), where people will send:

"cluster_link": {
}

in Request,

but response not.

- List has to return a response wrapped message anyway. So list is always inconsistent to these. IMHO, the buf style "must always be a request/response message" is great, but it's subjective.

@sago2k8 i think there were other reasons as well, we did initially do AIP style but moved to buf style for v1. I don't recall the other details anymore.

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+    int32 retry_backoff_ms = 8;
+    // Fetch request timeout
+    // If 0 is provided, defaults to 100ms
+    int32 fetch_wait_max_ms = 9;
```
I was about to suggest `google.protobuf.Duration` here, but it seems we are not using it anywhere so far. Not sure why

michael-redpanda replied:
We are using `google.protobuf.Duration` for intervals - we decided against it here because these translate directly to Kafka client properties which are already in `ms`.  We don't need anything more fine-grained than that.

rockwotj replied:
Basically these are kafka wire properties that are *only* at millisecond resolution. `google.protobuf.Duration` is nanosecond resolution, so I think having a `plain` int type makes more sense than the `Duration` possibly being rounded

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+message TopicMetadataSyncOptions {
+    // How often to sync metadata
+    // If 0 provided, defaults to 30 seconds
+    google.protobuf.Duration interval = 1;
```
Duration is used here though

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+        option (redpanda.pbgen.rpc) = {
+            authz: SUPERUSER
+        };
+    }
```
If the create is going to be a long running operation, we could use `google.longrunning.Operation`. In cloud we have our own flavor of operation which has proved to be useful for many APIs that return long running operations.

https://google.aip.dev/151

rockwotj replied:
it's not a long running operation

michael-redpanda replied:
As of right now it isn't - it should just be an update to the controller log

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+// Updates a cluster link
+message UpdateClusterLinkRequest {
+    // The cluster link to update
+    ClusterLink cluster_link = 1;
```
Based on our experience in cloud we've separated Create, and Update requests because they normally have different fields and different validations. For example, ID is not required in create but it is in Update.

In this case we'd do:

```proto
message UpdateClusterLinkRequest{
  message  ClusterLinkUpdate cluster_link = 1;
}

message UpdateClusterLinkRequest{
  message ClusterLinkCreate cluster_link = 1;
  google.protobuf.FieldMask update_mask = 2;
}

message ClusterLinkUpdate {
   // The name of the cluster link
    string name = 1 [(google.api.field_behavior) = REQUIRED];
    // The UUID of the cluster link
    string uid = 2 [
        (google.api.field_behavior) = OUTPUT_ONLY,
        (google.api.field_info).format = UUID4
    ];
    // Cluster link configuration
    ClusterLinkConfiguration configuration = 3;
}

message ClusterLinkCreate {
   // The name of the cluster link
    string name = 1 [(google.api.field_behavior) = REQUIRED];
    // Cluster link configuration
    ClusterLinkConfiguration configuration = 3;
}
```

michael-redpanda replied:
> For example, ID is not required in create but it is in Update.

I don't think that's true - the `uid` is an `OUTPUT_ONLY` field.

Our `Update` style is following [AIP-134](https://google.aip.dev/134)

simon0191 replied:
Yeah, this one is not AIP compliant, but we've found it helpful so we can have different validations for each of the methods, and be more explicit about the fields that are required in each method.

rockwotj replied:
We can't use protovalidate, so for us, it's more about conformance with cloud. Mike and I will discuss and come back.

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+option (redpanda.pbgen.cpp_namespace) = "proto::admin";
+
+service ClusterLinkService {
+    rpc CreateClusterLink(CreateClusterLinkRequest) returns (ClusterLink) {
```
In our APIs we use a dedicated type for Request and Response which buys a bit more flexibility. Pros and cons for each approach, but wanted to call it out.

```suggestion
    rpc CreateClusterLink(CreateClusterLinkRequest) returns (CreateClusterLinkResponse) {
```

michael-redpanda replied:
https://github.com/redpanda-data/redpanda/pull/27039#discussion_r2260681363

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+    repeated MirrorTopicStatus mirror_topic_status = 3;
+}
+// Status of a task
+message TaskStatus {
```
Is this TaskStatus specific for cluster links?
If yes, it should have a `ClusterLink` prefix like `ClusterLinkTaskStatus`, or it might be confusing because it's going to be a top level member of the admin namespace.

michael-redpanda replied:
gotcha sounds good!

birdayz replied:
It can also stay as TaskStatus and be moved inside ClusterLink message def (message defs can be nested)

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+/* Resources */
+// A ClusterLink resource
+message ClusterLink {
+    // The name of the cluster link
+    string name = 1 [(google.api.field_behavior) = REQUIRED];
+    // The UUID of the cluster link
+    string uid = 2 [
+        (google.api.field_behavior) = OUTPUT_ONLY,
+        (google.api.field_info).format = UUID4
+    ];
```
I find this a bit confusing. The name seems to be the main identifier. What is the UUID then needed for? Should we rename the field to `uuid` and possibly clarify in the comment?

michael-redpanda replied:
>  Should we rename the field to uuid and possibly clarify in the comment?

For UUIDs, I'm following [AIP-148](https://google.aip.dev/148#well-known-string-fields) which specifies the field name `uid`.

> The name seems to be the main identifier. What is the UUID then needed for?

Allows for distinguishing between cluster links that have the same name on different clusters.  It'll be part of the Client ID

rockwotj replied:
`uid` is from https://google.aip.dev/148#uid

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+}
+// Status of a MirrorTopic
+message MirrorTopicStatus {
+    string name = 1;
```
is this the cluster link name? if yes I suggest to be more explicit like `cluster_link_name`

rockwotj replied:
No it's the topic name. Maybe a comment would be good

michael-redpanda replied:
No it's the topic name - I can update the field or add a comment

michael-redpanda replied:
@rockwotj `<spider man meme>`

simon0191 commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+    // Cluster link is in the process of failing over
+    CLUSTER_LINK_STATE_FAILING_OVER = 3;
+    // Cluster link was failed over
+    CLUSTER_LINK_STATE_FAILED_OVER = 4;
```
Could there be an error state?

michael-redpanda replied:
possibly?  This hasn't been fully worked out yet so the state fields may change over time before the release.

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+        (google.api.field_info).format = UUID4
+    ];
+    // Cluster link configuration
+    ClusterLinkConfiguration configuration = 3;
```
```suggestion
    ClusterLinkConfiguration configurations = 3;
```

NIT: We prefer to go with plural for `configuration`. Established this as a standard based on docs guidelines which were the first standards.

rockwotj replied:
Same goes for the Message name in addition to the field name right?

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+// TLS settings
+message TLSSettings {
+    // May provide TLS as a set of files or PEM values
+    oneof tls_oneof {
```
The `_oneof` suffix is odd. Maybe just `oneof tls_credentials` or so?

michael-redpanda replied:
Yup will do

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+// Authentication config.  Currently only supporting SASL/SCRAM,
+// however made as a oneof for expansion
+message AuthenticationConfig {
+    oneof authentication_oneof {
```
`_oneof` suffix

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+// however made as a oneof for expansion
+message AuthenticationConfig {
+    oneof authentication_oneof {
+        ScramConfig scram_config = 1;
```
```suggestion
        ScramConfig scram_configuration = 1;
```

weeco commented on proto/redpanda/core/admin/cluster_link.proto:
```diff
@@ -0,0 +1,409 @@
+    // TLS settings
+    optional TLSSettings tls_settings = 4;
+    // Authentication settings
+    optional AuthenticationConfig authentication_config = 5;
```
```suggestion
    optional AuthenticationConfig authentication_configuration = 5;
```

michael-redpanda commented:
Force push:

* Addressed a number of PR comments
* Replaced `key_set` with `key_fingerprint`
* Added stubbed for using `timestamp` with SCRAM password - not yet supported so commented out
* Renamed ClusterLInk to ShadowLink
* Renamed MirrorTopic to ShadowTopic
* Moved stuff around

michael-redpanda commented:
Force push:

* Fix bazel lint error

michael-redpanda commented on proto/redpanda/core/admin/shadow_link.proto:
```diff
@@ -0,0 +1,427 @@
+import "google/api/field_info.proto";
+import "google/api/resource.proto";
+import "google/protobuf/duration.proto";
+// import "google/protobuf/timestamp.proto";
```
staged - will update once implemented

rockwotj replied:
https://github.com/redpanda-data/redpanda/pull/27171

michael-redpanda commented on proto/redpanda/core/admin/shadow_link.proto:
```diff
@@ -0,0 +1,427 @@
+    // Timestamp of when the password was last set - only valid if password_set
+    // is true
+    // google.protobuf.Timestamp password_set_at = 5
+    //    [(google.api.field_behavior) = OUTPUT_ONLY];
```
staged - will update once implemented

michael-redpanda commented:
Once https://github.com/redpanda-data/redpanda/pull/27171 merges, then I can use timestamps

michael-redpanda committed:
pbgen: Exclude google apis from being included in codegen

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push:

* Rebased off dev to get new pbgen updates

michael-redpanda commented:
Force push:

* Timestamp now supported, so use it for SCRAM credential set time
